### DISCUSSION TITLE
Carbon objects lose timezone after executing an action

### DISCUSSION THREAD

**Discussion by JeroenHauser** [livewire/livewire#8115] [7. Bugs (archived)]
### Livewire version

v3.4.8

### Laravel version

v10.48.2

### Which PHP version are you using?

PHP 8.3

### Steps To Reproduce

1. **Set a default timezone for the Laravel app in _.env_**
```
APP_TIMEZONE=Europe/Amsterdam
```
2. **Mount a Carbon object property when initialising the Livewire component**
```
class CarbonIssue extends Component
{

    public Carbon $date;

    public function mount(): void
    {
        $this->date = Carbon::make('March 14 2024 10:00:00');
    }

    public function nextMonth(): void
    {
        $this->date->addMonth();
    }

    public function render(): View
    {
        return view('livewire.carbon-issue');
    }

}
```

3. **Execute any action from the component interface**

```
<div>

    <button type="button" wire:click="nextMonth">
        Next month
    </button>

</div>
```
4. **Now the Carbon object property has lost its timezone setting**

**Initial Carbon object _before_ action**
Mind the `Europe/Amsterdam` on the last line
```
"date" => Illuminate\Support\Carbon @1713081600 {#2191
      #endOfTime: false
      #startOfTime: false
      #constructedObjectId: "000000000000088f0000000000000000"
      #localMonthsOverflow: null
      #localYearsOverflow: null
      #localStrictModeEnabled: null
      #localHumanDiffOptions: null
      #localToStringFormat: null
      #localSerializer: null
      #localMacros: null
      #localGenericMacros: null
      #localFormatFunction: null
      #localTranslator: null
      #dumpProperties: array:3 [
        0 => "date"
        1 => "timezone_type"
        2 => "timezone"
      ]
      #dumpLocale: null
      #dumpDateProperties: null
      date: 2024-03-14 00:00:00.0 Europe/Amsterdam (+01:00)
}
```
**Carbon object _after_ action** 
```
"date" => Illuminate\Support\Carbon @1713085200 {#951
      #endOfTime: false
      #startOfTime: false
      #constructedObjectId: "00000000000003b70000000000000000"
      #localMonthsOverflow: null
      #localYearsOverflow: null
      #localStrictModeEnabled: null
      #localHumanDiffOptions: null
      #localToStringFormat: null
      #localSerializer: null
      #localMacros: null
      #localGenericMacros: null
      #localFormatFunction: null
      #localTranslator: null
      #dumpProperties: array:3 [
        0 => "date"
        1 => "timezone_type"
        2 => "timezone"
      ]
      #dumpLocale: null
      #dumpDateProperties: null
      date: 2024-04-14 00:00:00.0 +01:00
}
```
5. **This is a problem because...**

While it seems to be of little impact (both Carbon objects use the `+01:00` UTC alteration), it actually conflicts with summer time. The initial Carbon object is aware of summer time starting om March 31 this year because it is aware of the timezone. The second Carbon object, however, is unaware of this.

This means that on dates after March 31 2024 (when initialised before that date), you get different timestamp outputs:
* When aware of timezone `Carbon::make('March 14 2024 10:00:00')->addMonth()->timestamp` results in `1713081600`
* When unaware of timezone `Carbon::make('March 14 2024 10:00:00')->addMonth()->timestamp` results in `1713085200`

That is a one hour difference, which is exactly the summer time hour.

6. **Proposed solution**

Always initialise Carbon object properties using the configured timezone, also after "reinitialising" them after executed actions.

### Repository or Wirebox URL

n/a, all code included in the report

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by joshhanley** [comment]
@JeroenHauser I had a quick look at this, and the reason this happens is because Livewire uses the ATOM format 

https://github.com/livewire/livewire/blob/c65b3f0798ab2c9338213ede3588c3cdf4e6fcc0/src/Mechanisms/HandleComponents/Synthesizers/CarbonSynth.php#L36

But if you have a look, Carbon doesn't have a serialisation string that include the timezone, instead they all just include the offset https://carbon.nesbot.com/docs/#api-commonformats

**Reply by JeroenHauser** [reply]
I see. Thanks for the clarification and pointing me in the right direction!

```
dd(
    \Carbon\Carbon::make('March 10 2024 10:00:00')->format(\DateTimeInterface::ATOM),
    // 2024-03-10T10:00:00+01:00
);
```

The `\DateTimeInterface::ATOM` format indeed adds the `+01:00` to the datetime string. After hydration, this results in a Carbon property that is unaware of the timezone.

```
dd(
    \Carbon\Carbon::make('March 10 2024 10:00:00')->format('Y-m-d H:i:s e'),
    // 2024-03-10 10:00:00 Europe/Amsterdam
);
```

Instead, I would suggest using a custom format that includes the timezone as shown above. This end up with a Carbon property that is still aware of the timezone after hydration.

```
dd(
    new \Carbon\Carbon(
        \Carbon\Carbon::make('March 10 2024 10:00:00')->format('Y-m-d H:i:s e'),
    ),
);

Carbon\Carbon @1710061200 {#802
  #endOfTime: false
  #startOfTime: false
  #constructedObjectId: "00000000000003220000000000000000"
  #localMonthsOverflow: null
  #localYearsOverflow: null
  #localStrictModeEnabled: null
  #localHumanDiffOptions: null
  #localToStringFormat: null
  #localSerializer: null
  #localMacros: null
  #localGenericMacros: null
  #localFormatFunction: null
  #localTranslator: null
  #dumpProperties: array:3 [
    0 => "date"
    1 => "timezone_type"
    2 => "timezone"
  ]
  #dumpLocale: null
  #dumpDateProperties: null
  date: 2024-03-10 10:00:00.0 Europe/Amsterdam (+01:00)
}
```

At the moment, I cannot think of any side effects, since after interpreting the timezone, the Carbon object property is still aware of the `+01:00` UTC alteration.

**Reply by joshhanley** [reply]
@JeroenHauser the main issue I can see now is if people depend on the current format being available to the front end, then this will be a breaking change. I wonder whether we can add timezone to the meta data in the synth (second parameter in the array returned from the dehydrate method) instead so it can be used to restore the property but doesn't change what is sent to the front end. That way the Carbon instance is still correct after hydration.

**Reply by JeroenHauser** [reply]
Hi @joshhanley, any update on this? Just checking in, no pressure.

For now, I'm using a workaround reinitialising the Carbon object myself:
```
public function nextMonth(): void
{
    $this->date = Carbon::make($this->date->format('Y-m-d H:i:s'));
    $this->date->addMonth();
}
```

The new Carbon object uses the default timezone again.

**Reply by joshhanley** [reply]
@JeroenHauser it's not something I'm working on, but PRs are welcome! 🙂 Otherwise you can implement your own carbon Synth, which will take priority over the inbuilt carbon Synth. See here for how to build synths (plus you can look at the built in one) https://livewire.laravel.com/docs/synthesizers
